File: CSharp/LeetCode/004-MedianOfTwoSortedArrays.cs
namespace LeetCode
{
    public class _004_MedianOfTwoSortedArrays
    {
        public double FindMedianSortedArrays(int[] nums1, int[] nums2)
        {
            return FindKth(nums1, 0, nums2, 0, (nums1.Length + nums2.Length) / 2);
        }

        double FindKth(int[] nums1, int startIndex1, int[] nums2, int startIndex2, int k)
        {
            if (nums1.Length - startIndex1 < nums2.Length - startIndex2) { return FindKth(nums2, startIndex2, nums1, startIndex1, k); }
            if (nums2.Length <= startIndex2) { return nums1[k]; }
            if (k == 1) { return nums1[startIndex1] < nums2[startIndex2] ? nums1[startIndex1] : nums2[startIndex2]; }

            var index1 = k / 2 + startIndex1 < nums1.Length - startIndex1 ? k / 2 : nums1.Length - startIndex1;
            var index2 = k - index1 + startIndex2;
            if (nums1[index1] > nums2[index2])
            {
                return FindKth(nums1, startIndex1, nums2, index2, k - index2 + startIndex2);
            }
            else if (nums1[index1] < nums2[index2])
            {
                return FindKth(nums1, index1, nums2, startIndex2, k - index1 + startIndex1);
            }
            else
            {
                return nums1[index1 - 1];
            }
        }
    }
}

File: C++/LeetCode/004-MedianOfTwoSortedArrays.cpp
#include "stdafx.h"
#include "004-MedianOfTwoSortedArrays.h"


_004_MedianOfTwoSortedArrays::_004_MedianOfTwoSortedArrays()
{
}


_004_MedianOfTwoSortedArrays::~_004_MedianOfTwoSortedArrays()
{
}

int findKth(vector<int> & nums1, int startIndex1, vector<int> & nums2, int startIndex2, int k) {
	int nums1Left = (int)nums1.size() - startIndex1;
	int nums2Left = (int)nums2.size() - startIndex2;

	if (nums1Left < nums2Left) { return findKth(nums2, startIndex2, nums1, startIndex1, k); }
	if (nums2Left <= 0) { return nums1[startIndex1 + k - 1]; }
	if (k == 1) { return min(nums1[startIndex1], nums2[startIndex2]); }

	int index1 = min(k / 2, nums1Left);
	int index2 = min(k - index1, nums2Left);
	if (nums1[startIndex1 + index1 - 1] > nums2[startIndex2 + index2 - 1])
	{
		return findKth(nums1, startIndex1, nums2, startIndex2 + index2, k - index2);
	}
	else if (nums1[startIndex1 + index1 - 1] < nums2[startIndex2 + index2 - 1])
	{
		return findKth(nums1, startIndex1 + index1, nums2, startIndex2, k - index1);
	}
	else
	{
		return nums1[startIndex1 + index1 - 1];
	}
}

double _004_MedianOfTwoSortedArrays::findMedianSortedArrays(vector<int> & nums1, vector<int> & nums2)
{
	int m = (int)nums1.size() + (int)nums2.size();

	if (m & 0x1)
	{
		return findKth(nums1, 0, nums2, 0, (m + 1) / 2);
	}
	else
	{
		return (findKth(nums1, 0, nums2, 0, m / 2) + findKth(nums1, 0, nums2, 0, m / 2 + 1)) / 2.0;
	}
}
